a3p-integration/proposals/p:upgrade-19/provisionPool.test.js

```diff
@@ -0,0 +1,230 @@
+/* eslint-env node */
+/**
+ * @file The goal of this file is to make sure v28-provisionPool and v14-bank can be successfully
+ * upgraded. These vats are related because of the issues below;
+ * - https://github.com/Agoric/agoric-sdk/issues/8722
+ * - https://github.com/Agoric/agoric-sdk/issues/8724
+ *
+ * The test scenario is as follows;
+ * 1. Upgrade provisionPool. This upgrade overrides provisionWalletBridgerManager with a durable one
+ * 2. Add a new account and successfully provision it
+ *   - Observe new account's address under `published.wallet.${address}`
+ * 3. Send some USDC_axl to provisionPoolAddress and observe its IST balances increases accordingly
+ * 4. Introduce a new asset to the chain and start a PSM instance for the new asset
+ *   4a. Deposit some of that asset to provisionPoolAddress
+ *   4b. Observe provisionPoolAddress' IST balance increase by the amount deposited in step 4a
+ * 5. Perform a null upgrade for provisionPool. This upgrade does NOT override provisionWalletBridgerManager
+ *   - The goal here is to allow testing the bridgeHandler from the first upgrade is in fact durable
+ * 6. Auto provision
+ *   6a. Introduce a new account
+ *   6b. Fund it with IST and ATOM to be able to open a vault
+ *   6c. Try to open a vault WITHOUT provisioning the newly introduced account
+ *   6d. Observe the new account's address under `published.wallet`
+ * 7. Same as step 2. Checks manual provision works after null upgrade
+ */
+
+import '@endo/init';
+import test from 'ava';
+import { execFileSync } from 'node:child_process';
+import {
+  addUser,
+  makeAgd,
+  evalBundles,
+  agd as agdAmbient,
+  agoric,
+  getISTBalance,
+  getDetailsMatchingVats,
+  VALIDATORADDR,
+  CHAINID,
+  GOV1ADDR,
+  openVault,
+  ATOM_DENOM,
+} from '@agoric/synthetic-chain';
+import {
+  makeVstorageKit,
+  retryUntilCondition,
+  waitUntilAccountFunded,
+  waitUntilContractDeployed,
+} from '@agoric/client-utils';
+import { NonNullish } from '@agoric/internal';
+
+const PROVISIONING_POOL_ADDR = 'agoric1megzytg65cyrgzs6fvzxgrcqvwwl7ugpt62346';
+
+const ADD_PSM_DIR = 'addUsdLemons';
+const DEPOSIT_USD_LEMONS_DIR = 'depositUSD-LEMONS';
+const UPGRADE_PP_DIR = 'upgradeProvisionPool';
+const NULL_UPGRADE_PP_DIR = 'nullUpgradePP';
+
+const USDC_DENOM = NonNullish(process.env.USDC_DENOM);
+
+const agd = makeAgd({ execFileSync }).withOpts({ keyringBackend: 'test' });
+
+const ambientAuthority = {
+  query: agdAmbient.query,
+  follow: agoric.follow,
+  setTimeout,
+  log: console.log,
+};
+
+/**
+ * @param {string} addr
+ * @param {string} wanted
+ * @param {string} [from]
+ */
+export const bankSend = (addr, wanted, from = VALIDATORADDR) => {
```

Comment:
please pull `bankSend`, `provision`, `introduceAndProvision`, `getProvisionedAddress`,  and `checkUserProvisioned` out to a separate library file that can be shared. We shouldn't  have to define these more than once.

Maybe `getProvisionedAddress` is simple enough that it doesn't have to go in the library.

Reply:
> please pull `bankSend`, `provision`, `introduceAndProvision`, `getProvisionedAddress`, and `checkUserProvisioned` out to a separate library file that can be shared. We shouldn't have to define these more than once.

https://github.com/Agoric/agoric-sdk/issues/10514 already tracks this one. I can put those methods into a helper module then move them to the shared library as part of #10514. I think we shouldn't wait for the shared library to land this PR.

Reply:
> I think we shouldn't wait for the shared library to land this PR
That's fine.